### PR TITLE
[Gdpr] Amend TCF v2 purpose consent logic

### PR DESCRIPTION
**Pull request by vguerci**

### CODE REVIEW

File: CriteoPublisherSdk/Tests/UnitTests/DataConsent/CR_GdprTests.m [hunk 1]
```diff
@@ -162,6 +162,46 @@ - (void)testPurposeConsentMissingWithConsentStringForTcf2_0 {
...
+    [self.userDefaults clearGdpr];                                                     \
+    [self.userDefaults setGdprTcf##tcfVersion##GdprApplies:@YES];                      \
+    [self.userDefaults setGdprTcf2_0PublisherRestrictions:restrictions forPurpose:id]; \
+    XCTAssertEqual([self.gdpr publisherRestrictionsForPurpose:id], expected);          \
+  } while (0);
+
+  AssertPublisherRestrictions(1_1, nil, 1, CR_GdprTcfPublisherRestrictionTypeNone);
+  AssertPublisherRestrictions(2_0, @"", 1, CR_GdprTcfPublisherRestrictionTypeNone);
+  AssertPublisherRestrictions(2_0, @"too short", 1, CR_GdprTcfPublisherRestrictionTypeNone);
+  AssertPublisherRestrictions(2_0, ([NSString stringWithFormat:@"%091d", 0]), 1,
```
Comment (TurpIF) [inline CriteoPublisherSdk/Tests/UnitTests/DataConsent/CR_GdprTests.m:179]: nice trick

File: CriteoPublisherSdk/Tests/Utility/Categories/NSUserDefaults+GDPR.m [hunk 2]
```diff
@@ -27,6 +27,11 @@ - (void)clearGdpr {
   [self removeObjectForKey:NSString.gdprAppliesUserDefaultsKeyTcf1_1];
   [self removeObjectForKey:NSString.gdprConsentStringUserDefaultsKeyTcf1_1];
   [self removeObjectForKey:NSString.gdprPurposeConsentsStringForTcf2_0];
+  for (int id = 0; id <= 10; ++id) {
```
Comment (TurpIF) [inline CriteoPublisherSdk/Tests/Utility/Categories/NSUserDefaults+GDPR.m:30]: Where does this 10 comes from ? What about adding a comment or extracting it in a named constant ?

File: CriteoPublisherSdk/Tests/UnitTests/DataConsent/CR_GdprTests.m [hunk 3]
```diff
@@ -202,6 +202,46 @@ - (void)testPublisherRestrictionsMWithConsentStringForTcf2_0 {
...
+    [self.userDefaults setGdprTcf2_0VendorConsents:consents];     \
+    XCTAssertEqual([self.gdpr isVendorConsentGiven], expected);   \
+  } while (0);
+
+  AssertVendorConsent(1_1, nil, YES);
+  AssertVendorConsent(2_0, @"", YES);
+  AssertVendorConsent(2_0, @"malformed", YES);
+  AssertVendorConsent(
+      2_0,
+      @"---------+---------+---------+---------+---------+---------+---------+---------+---------+X--------+",
```
Comment (TurpIF) [inline CriteoPublisherSdk/Tests/UnitTests/DataConsent/CR_GdprTests.m:221]: Isn't it possible to execute the trick you've done in the commit before ? Formatting an int on `¨%091d` ?

File: CriteoPublisherSdk/Sources/DataConsent/CR_Gdpr.h [hunk 4]
```diff
@@ -65,6 +77,14 @@ typedef NS_ENUM(NSInteger, CR_GdprTcfVersion) {
  */
 - (BOOL)isConsentGivenForPurpose:(NSUInteger)id;
 
+/**
+ * Gives the Criteo publisher restrictions for a given purpose
+ *
+ * @param id Purpose id
+ * @return Publisher Restriction type
+ */
+- (CR_GdprTcfPublisherRestrictionType)publisherRestrictionsForPurpose:(NSUInteger)id;
```
Comment (TurpIF) [inline CriteoPublisherSdk/Sources/DataConsent/CR_Gdpr.h:86]: AFAIU, we're only using this method with id=1. Is there any reason why this is parameterized and not hardcoded ?

File: CriteoPublisherSdk/Sources/DataConsent/CR_GdprVersion.h [hunk 5]
```diff
@@ -61,6 +61,30 @@ extern NSString *const CR_GdprConsentStringForTcf1_1Key;
  */
```
Comment (Amokrane) [inline CriteoPublisherSdk/Sources/DataConsent/CR_GdprVersion.h:61]: I am curious what is the difference between `CR_GdprVersion` and `CR_Gdpr`? I see that both declare the same methods, such as: `isVendorConsentGiven` and `hasVendorLegitimateInterest `

File: CriteoPublisherSdk/Sources/DataConsent/CR_GdprVersion.m [hunk 6]
```diff
@@ -20,9 +20,13 @@
 #import "CR_GdprVersion.h"
 #import "NSString+Tcf.h"
 
+NSUInteger const CR_CriteoTcfVendorID = 91;
```
Comment (Amokrane) [inline CriteoPublisherSdk/Sources/DataConsent/CR_GdprVersion.m:23]: I suggest adding a comment that explains where `91` comes from as it's not trivial.

File: CriteoPublisherSdk/Sources/DataConsent/CR_GdprVersion.m [hunk 7]
```diff
@@ -103,6 +122,54 @@ - (BOOL)isConsentGivenForPurpose:(NSUInteger)id {
...
 }
 
+- (CR_GdprTcfPublisherRestrictionType)publisherRestrictionsForPurpose:(NSUInteger)id {
+  if (self.publisherRestrictionsKeyFormat == nil) {
+    return CR_GdprTcfPublisherRestrictionTypeNone;
+  }
+  NSString *purposeRestrictionKey =
+      [NSString stringWithFormat:self.publisherRestrictionsKeyFormat, id];
+  NSString *restrictionsForPurpose = [self.userDefaults stringForKey:purposeRestrictionKey];
+  NSUInteger criteoIndex = CR_CriteoTcfVendorID - 1;
```
Comment (Amokrane) [inline CriteoPublisherSdk/Sources/DataConsent/CR_GdprVersion.m:132]: By including the offset in the constant initialization, you could save this line (unless the non-offset value is needed somewhere else?)

**Review by Amokrane — CHANGES_REQUESTED** [review]

**Review by Amokrane — CHANGES_REQUESTED** [review]
minor requests.

File: CriteoPublisherSdk/Sources/DataConsent/CR_Gdpr.h (hunk 4 above)
Comment (TurpIF) [reply CriteoPublisherSdk/Sources/DataConsent/CR_Gdpr.h:86]: Nvm, I saw that you were using purpose=7 before.

File: CriteoPublisherSdk/Tests/UnitTests/DataConsent/CR_GdprTests.m (hunk 3 above)
Comment (vguerci) [reply CriteoPublisherSdk/Tests/UnitTests/DataConsent/CR_GdprTests.m:221]: It is, it is just that I didn't want to trick myself with tricks everywhere and wanted to be explicit at least once :)
I'll edit this.

File: CriteoPublisherSdk/Sources/DataConsent/CR_GdprVersion.h (hunk 5 above)
Comment (vguerci) [reply CriteoPublisherSdk/Sources/DataConsent/CR_GdprVersion.h:61]: This is how TCF v1 & v2 overlap has been handled. One implementation per version. Switching logic being on top.

File: CriteoPublisherSdk/Sources/DataConsent/CR_GdprVersion.m (hunk 6 above)
Comment (vguerci) [reply CriteoPublisherSdk/Sources/DataConsent/CR_GdprVersion.m:23]: Done.

File: CriteoPublisherSdk/Sources/DataConsent/CR_GdprVersion.m (hunk 7 above)
Comment (vguerci) [reply CriteoPublisherSdk/Sources/DataConsent/CR_GdprVersion.m:132]: Thanks! That comment made me realise that there was an off-by-one issue on the particular edge case checked there.

Fixed, which makes that variable to be removed.

File: CriteoPublisherSdk/Tests/Utility/Categories/NSUserDefaults+GDPR.m (hunk 2 above)
Comment (vguerci) [reply CriteoPublisherSdk/Tests/Utility/Categories/NSUserDefaults+GDPR.m]: Done

File: CriteoPublisherSdk/Tests/UnitTests/DataConsent/CR_GdprTests.m (hunk 3 above)
Comment (vguerci) [reply CriteoPublisherSdk/Tests/UnitTests/DataConsent/CR_GdprTests.m]: Oh also it was because you cannot apply this trick to `X` but nevermind ;p

**Review by TurpIF — DISMISSED** [review]
LGTM

File: CriteoPublisherSdk/Sources/Feedback/CR_FeedbackConsentGuard.m [hunk 8]
```diff
@@ -70,11 +70,27 @@ - (void)sendFeedbackBatch {
...
+  CR_Gdpr *gdpr = _consent.gdpr;
+  BOOL hasUserConsent = [gdpr isConsentGivenForPurpose:1];
+  CR_GdprTcfPublisherRestrictionType publisherRestriction =
+      [gdpr publisherRestrictionsForPurpose:1];
+  BOOL hasPublisherRestrictions =
+      publisherRestriction == CR_GdprTcfPublisherRestrictionTypeNotAllowed ||
+      publisherRestriction == CR_GdprTcfPublisherRestrictionTypeRequireLegitimateInterest;
+  BOOL hasVendorConsent = [gdpr isVendorConsentGiven];
+  BOOL hasVendorLegitimateInterest = [gdpr hasVendorLegitimateInterest];
+  return hasUserConsent && !hasPublisherRestrictions && hasVendorConsent &&
```
Comment (Amokrane) [inline CriteoPublisherSdk/Sources/Feedback/CR_FeedbackConsentGuard.m:92]: Privacy question: if `requiredLegitimateInterest` is set to `true`, and the vendor has legitimate interest established, don't we have the right to store?

Comment (vguerci) [reply CriteoPublisherSdk/Sources/Feedback/CR_FeedbackConsentGuard.m:92]: Not with my understanding of our internal policy, which are written so the "Vendor" consent does not override "User" consent.

**Review by Amokrane — APPROVED** [review]

**Comment by vguerci** [comment]
Thank you! 👍
